### PR TITLE
Add log data model

### PR DESCRIPTION
**Pull request by tigrannajaryan**
The content of data-model.md is primarily copied from OTEP 0097.
I removed sections that were only relevant in the context of OTEP
discussion and made minimal changes so that this is no longer
a "proposal" but a "specification".

### CODE REVIEW

File: specification/logs/data-model.md [hunk 1]
```diff
@@ -0,0 +1,1255 @@
...
+SeverityText   |The severity text (also known as log level).
+SeverityNumber |Numerical value of the severity.
+Name           |Short event identifier.
+Body           |The body of the log record.
+Resource       |Describes the source of the log.
+Attributes     |Additional information about the event.
+
+Below is the detailed description of each field.
+
+### Field: `Timestamp`
```
Comment (zenmoto) [inline specification/logs/data-model.md:165]: We should probably be explicit as to whether this should be set if the record is received but not parsed. For example if a docker logging driver is outputting in this format, the time that it receives the message won't be the same as the time that the event was generated, but it may not have the capability to parse the date from the log text. In that case, this entry should remain empty, right? It might be worth calling that out.

File: specification/logs/data-model.md [hunk 2]
```diff
@@ -0,0 +1,1255 @@
...
+specification. At the time of writing the specification defines one flag - the
+SAMPLED flag. This field is optional.
+
+### Severity Fields
+
+#### Field: `SeverityText`
+
+Type: string.
+
+Description: severity text (also known as log level). This is the original
```
Comment (zenmoto) [inline specification/logs/data-model.md:206]: Is the preference to add the default SeverityText, or to leave it empty and map from the number to the text when exporting?

File: specification/logs/data-model.md [hunk 3]
```diff
@@ -0,0 +1,1255 @@
...
+capitalization or abbreviated, e.g. "Info" vs "Information").
+
+#### Comparing Severity
+
+In the contexts where severity participates in less-than / greater-than
+comparisons `SeverityNumber` field should be used. `SeverityNumber` can be
+compared to another `SeverityNumber` or to numbers in the 1..24 range (or to the
+corresponding short names).
+
+When severity is used in equality or inequality comparisons (for example in
```
Comment (zenmoto) [inline specification/logs/data-model.md:357]: I'm not sure I agree with this- it seems like the advantage of having a `SeverityNumber` and these mappings is that we can treat `SeverityNumber=7, SeverityText=Cucumber` as semantically equal to `SeverityNumber=7, SeverityText=Melon` as the default case, while still allowing users to specify equality on the `SeverityText` if their use-case demands it.

File: specification/logs/data-model.md [hunk 4]
```diff
@@ -0,0 +1,1255 @@
...
+with `SeverityText` field equal to "Informational" and `SeverityNumber` field
+equal to INFO then it may be preferable from the user experience perspective to
+ensure that **severity="Informational"** and **severity="INFO"** conditions both
+to are TRUE for that record.
+
+### Field: `Name`
+
+Type: string.
+
+Description: Short event identifier that does not contain varying parts.
```
Comment (zenmoto) [inline specification/logs/data-model.md:370]: My understanding here is that this name would also be where we would put the logger name from a logging library, correct? In Java this is often a fully-qualified class name, which may often exceed 50 characters.

File: specification/logs/data-model.md [hunk 5]
```diff
@@ -0,0 +1,1255 @@
...
+    <td>index</td>
+    <td>string</td>
+    <td>The name of the index by which the event data is to be indexed. The index you specify here must be within the list of allowed indexes if the token has the indexes parameter set.</td>
+    <td>TBD, most like will go to attributes</td>
+  </tr>
+</table>
+
+### Log4j
+
+<table>
```
Comment (zenmoto) [inline specification/logs/data-model.md:709]: We should probably add the mapping for the Logger attribute for Log4j as it's pretty central to that and similar libraries.

File: specification/logs/data-model.md (hunk 4 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:370]: The `Name` describes what happened. Logger name describes where it happened. So Logger name fits better in the `Resource` or alternatively can be an `Attribute`.

See also this debate about InstrumentationLibrary: https://github.com/open-telemetry/opentelemetry-proto/pull/150
Since I lost the debate and we kept InstrumentationLibrary I think the best fit for Logger name is InstrumentationLibrary.name.

File: specification/logs/data-model.md (hunk 5 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:709]: According to this mapping it will go to `Attributes`, which I think is not wrong (but see also my other [comment](https://github.com/open-telemetry/opentelemetry-specification/pull/694/files#r451828805)). We will need to come up with the name of the attribute and add it to our semantic conventions.

File: specification/logs/data-model.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:165]: I am not sure what is the right approach in that case. Whoever is receiving the log should make a decision: is it better to set the timestamp to the time of reception or keep it empty and let the parser fill it later (which may not happen if there is no parser).
I don't know what's the best approach, so I think it is best left out and let whoever implements the receiver decide what's best for their particular use case.

File: specification/logs/data-model.md (hunk 2 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:206]: I think if in the future we decide to implement OpenTelemetry Logging API we can leave SeverityText empty. For all other cases (which is really all the cases we intend to support for now) I think it is best to set SeverityText explicitly.

File: specification/logs/data-model.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:357]: >we can treat SeverityNumber=7, SeverityText=Cucumber as semantically equal to SeverityNumber=7, SeverityText=Melon

Yes, that's the intent.

Sorry, I do not quite understand what you disagree with. Please clarify.

File: specification/logs/data-model.md (hunk 4 above)
Comment (zenmoto) [reply specification/logs/data-model.md:370]: `Resource` is probably a bad place for it, as one application will often have a number of loggers (with Java being an extreme case, often having one for each class), so it probably belongs on the event. For many logging libraries it is going to be an attribute on every event- it seems like something that should be hoisted up from the attributes? Somehow I'd associated `name` with the logger name until now (in fact the log4j libraries I recently added put the logger name there).

File: specification/logs/data-model.md (hunk 3 above)
Comment (zenmoto) [reply specification/logs/data-model.md:357]: I think that I'm simply misunderstanding what you are intending in this section.

Comment (tigrannajaryan) [reply specification/logs/data-model.md:357]: Let me try to clarify.

Let's imagine my Logging product has a UI where I can filter log records by fields in this data model. Let's imagine my filter allows me to write a filter condition `SeverityText=="ERROR"`. Which log record should this filter match?

Obviously it has to match the records that have "ERROR" as the precise value of `SeverityText`. I am proposing that in addition to that it also matches any log record for which `SeverityNumber` is equal to 17 (which is the numeric value for "ERROR") _regardless_ of what the `SeverityText` of that log record is.

This way I can implement by `SeverityText` filter as a dropdown which contains the limited list of elements that are defined in this data model (the 24 levels that we defined above) and have a reasonable behavior for the cases when `SeverityText` is some custom text that does not match what this document defines for the `SeverityNumber` recorded.

Does this make sense or you think it is an unnecessary recommendation?

File: specification/logs/data-model.md (hunk 4 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:370]: We have had this discussion (whether the logger or library name belongs to the Resource) and did not come to a definitive conclusion.

Perhaps you can help with some fresh ideas. Why do you think the number of loggers make the Resource a bad place for recording the logger name?

Our definition of the Resource is that it is the location where the event happened. What makes Logger name unfit for the definition?

File: specification/logs/data-model.md (hunk 1 above)
Comment (zenmoto) [reply specification/logs/data-model.md:165]: I think that it should be explicit, though there are tradeoffs either way. Extracting the date from the log entry is both costly and error-prone- if we can rely that if the time exists on the event it can be relied upon to be the event time, we can entirely avoid that step. On the other hand, we risk not having a timestamp at all if we do that, but that means we need to try to parse the date whether we have a timestamp or not, and this nicely consistent value only works as a fallback if we want the true event time. Perhaps some sort of flag saying which type of time it is would be best, but in lieu of that, being specific about the approach would help keep the semantics consistent downstream.

File: specification/logs/data-model.md (hunk 3 above)
Comment (zenmoto) [reply specification/logs/data-model.md:357]: I think that's probably best left to the UX designers to figure out as long as we are clear where those values come from and what they signify.

File: specification/logs/data-model.md (hunk 4 above)
Comment (zenmoto) [reply specification/logs/data-model.md:370]: My understanding of the resource is that it should remain consistent for the lifetime of the program/pod/instance. If we look at the set of key-value pairs in a resource as being a unique resource, that changes a running program instance from being one resource and turns it into many, sometimes hundreds. From a conceptual standpoint, the resource stops being a serverless function, and instead becomes the error handler for the HTTP client being used by that serverless function, which becomes I think less useful.

From a functional standpoint it also becomes more difficult to transmit that information- if using existing log formats, we may largely choose to forego resource information in lieu of a simple proxy like hostname or container ID. In that case we have no place to put the logger name. Also it means that if in OTLP we send the resource once per batch, now we need to batch per logger, which is going to greatly complicate the batching procedures. Since a single request undoubtedly cuts through many classes, the cardinality on that division is also going to be very high, so that's going to greatly decrease the efficiency of the batching.

Finally it seems that the resource is designed to be pretty rarely changed- at least in the Java implementation, it's implemented in an immutable way, and I'm not sure what the performance penalty would be for creating a `Resource` instance for each attribute- it certainly feels counter to the intent, but I could certainly be wrong there.

File: specification/logs/data-model.md (hunk 5 above)
Comment (zenmoto) [reply specification/logs/data-model.md:709]: This is something that's used in most logging libraries, and for those that do use it, it's added to 100% of the events, so it seems like the cost of adding the key name each time could be significant.

Comment (zenmoto) [reply specification/logs/data-model.md:709]: I apologize for misunderstanding the intent of the `Name` attribute in earlier iterations, and so to be bringing it up so late in the process.

File: specification/logs/data-model.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:165]: I think the specification defines which type of timestamp it is:
>Time when the event occurred measured by the origin clock.

If the timestamp is known we put it here. If not known we leave it empty. If later someone is capable of extracting the timestamp from the body they can do it and put it in this field. Alternatively they can put the time or arriving as the best approximation of time of event if that's the best they can do. Do we need to be more prescriptive than we already are with this field? Perhaps you can point me to some prior art, a log data format which does what you describe?

Maybe we just need to reword this as:
>Time when the event occurred (or best approximation of it) measured by the origin clock.

File: specification/logs/data-model.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:357]: OK, I think that's fair. I can remove this part.

Comment (tigrannajaryan) [reply specification/logs/data-model.md]: Removed.

File: specification/logs/data-model.md (hunk 4 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:361]: > My understanding of the resource is that it should remain consistent for the lifetime of the program/pod/instance.

The spec defines the Resource as:

>A Resource is an immutable representation of the entity producing telemetry.

A program/pod/instance is indeed one possible type of a Resource, but I don't think we prohibit smaller parts of a program to be a Resource. I do not insist that we should put the logger name in the Resource, perhaps an Attribute  is a better place, but Resource definition does not prohibit the usage I was suggesting.


> If we look at the set of key-value pairs in a resource as being a unique resource, that changes a running program instance from being one resource and turns it into many, sometimes hundreds. From a conceptual standpoint, the resource stops being a serverless function, and instead becomes the error handler for the HTTP client being used by that serverless function, which becomes I think less useful.

I do not think a Resource strictly corresponds to a running program instance. Consider logs collected from a host, e.g. from a "syslog". The log entries will be about many programs, however I can indeed call the host a Resource.

Perhaps what you propose is valid and we need to have a stricter definition of a Resource, but that's not how it is defined today. If you want it to be changed I think it needs to be a separate OTEP or spec PR.

> From a functional standpoint it also becomes more difficult to transmit that information- if using existing log formats, we may largely choose to forego resource information in lieu of a simple proxy like hostname or container ID. In that case we have no place to put the logger name. Also it means that if in OTLP we send the resource once per batch, now we need to batch per logger, which is going to greatly complicate the batching procedures. Since a single request undoubtedly cuts through many classes, the cardinality on that division is also going to be very high, so that's going to greatly decrease the efficiency of the batching.

This is a good point. If we want to avoid batching per logger then indeed putting the logger name in the Attributes is a better approach. There is a cost in transmitting it for every event to which I offered a solution in https://github.com/open-telemetry/opentelemetry-proto/pull/150#issuecomment-646331469 but it was not accepted at the time it was discussed.

File: specification/logs/data-model.md (hunk 5 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:700]: No problem. I don't think spec should be set in stone, now is a good time to have these discussions, we are still in Alpha.

File: specification/logs/data-model.md (hunk 1 above)
Comment (zenmoto) [reply specification/logs/data-model.md:165]: I think maybe I'm just splitting a hair that's fine intact. I think that where I'm coming from originates with problems I've seen in Splunk installations where we have a few timestamps- primary is _time which is set from parsing the timestamp if we can find and parse it in the event, and receipt time if not. We also have indextime, which is the time that the event actually got indexed. Sometimes it's possible to have a significant difference between _time and indextime, and sometimes there's a significant difference between _time and the human-readable time in the event. How those deltas get there varies (and is sometimes a good troubleshooting tool), but it seems to almost universally cause confusion.

File: specification/logs/data-model.md (hunk 4 above)
Comment (zenmoto) [reply specification/logs/data-model.md:361]: You are right that a program instance isn't a good description of a resource, but I think that it describes the intended granularity- something that's running somewhere. I think that we could describe individual classes as resources, but to be useful I think the first step a user would need to take would be to aggregate if we did that- grouping by resource wouldn't give a reasonable view of a request or event, it would have to be a grouping by an aggregate of resources that would be somewhat awkward.

The logger is also modeled as just an attribute of a log event in most logging libraries- for example [Log4j here](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/LogEvent.html). It's used internally in the library in a way that's somewhat analogous to a hierarchical tag, where the logger is created with a tag to apply to all events that are added to it, and then filters and mappings are applied to the event depending on the tag that it has.

File: specification/logs/data-model.md (hunk 5 above)
Comment (zenmoto) [reply specification/logs/data-model.md:700]: Resolving this one, since it's a sidebar on the discussion above.

File: specification/logs/data-model.md (hunk 4 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:361]: Sounds good. Let's go with an Attribute for logger name.

We need a semantic convention for. We discussed a while back adopting ECS-like conventions. If we go with that we will use "log.logger" attribute for logger name.

File: specification/logs/data-model.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:165]: Do you propose to prohibit recording anything but the original timestamp in this field? Or do you suggest to have an additional timestamp field for non-origin timestamp? This was discussed during the OTEP and we decided to put the additional timestamps in the Attributes if needed.

Comment (zenmoto) [reply specification/logs/data-model.md:165]: Actually, I think that’s perfect. If we can make this explicitly as event time on origin origin clock and determine an attribute for receipt time the semantics are very clear.

File: specification/logs/data-model.md (hunk 4 above)
Comment (zenmoto) [reply specification/logs/data-model.md:361]: I can live with that. I've got concern about adding 10k+ bytes per 1k events over a field, but we can prove it out with this and adjust later if needed.

Comment (tigrannajaryan) [reply specification/logs/data-model.md:361]: Resolving this, I assume we are good, feel free to reopen.

File: specification/logs/data-model.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/logs/data-model.md:165]: In that case I am not sure what is missing. We already say this:
>Time when the event occurred measured by the origin clock.

Is this different from what you expect: 
>make this explicitly as event time on origin origin clock

As for defining an attribute for receipt time, sure we can do that in a separate PR. That will need to be filed against semantic conventions docs.

Comment (zenmoto) [reply specification/logs/data-model.md:165]: No, I think that’s explicit enough. When we have that convention I would like to link it here, but I think it’s good until then.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers @open-telemetry/specs-logs-approvers please review.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers this is merely adding a previously accepted OTEP to the specification repo with minimal edits. Please review and approve.
